src/main/scala/com/webtrends/harness/component/akkahttp/websocket/AkkaHttpWebsocket.scala

```diff
@@ -16,261 +16,97 @@
-//          case Failure(_) => bean.addValue(key, uri)
-//        }
-//        true
-//
-//      // Treat the value as a string
-//      case (uri, test) if test.head == '%' =>
-//        bean.addValue(test.drop(1), uri)
-//        true
-//
-//      // Only match if the value is an INT
-//      case (uri, test) if test.head == '#' =>
-//        Try(uri.toInt) match {
-//          case Success(v) =>
-//            bean.addValue(test.drop(1), v.asInstanceOf[Integer])
-//            true
-//          case Failure(_) =>
-//            false
-//        }
-//
-//      case (uri, test) =>
-//        test.toLowerCase.split('|').contains(uri.toLowerCase)
-//    }
-//
-//    if (matched) Some(bean)
-//    else None
-//  }
-//}
\ No newline at end of file
+  def chosenCompression(headers: Map[String, String]): Option[CompressionType] = {
+    headers.map(h => h._1.toLowerCase -> h._2).get("accept-encoding").flatMap { encoding =>
+      val encSet = encoding.split(",").map(_.trim).toSet
+      val selected = supportedCompression.keySet.intersect(encSet).headOption
+      selected.map(algo => CompressionType(algo, supportedCompression(algo)))
+    }
+  }
+}
+
+class AkkaHttpWebsocket[I: ClassTag, O <: Product : ClassTag, A <: Product : ClassTag](
+  authHolder: A,
+  inputHandler: (A, TextMessage) => Future[I],
+  businessLogic: I => Future[O],
+  responseHandler: O => TextMessage,
+  onClose: A => Unit = {_: A => ()},
+  errorHandler: PartialFunction[Throwable, Directive] = AkkaHttpEndpointRegistration.wsErrorDefaultHandler,
+  options: EndpointOptions = EndpointOptions.default)(implicit ec: ExecutionContext, mat: Materializer) extends LoggingAdapter {
+  var closed: AtomicBoolean = new AtomicBoolean(false)
+
+  // This the the main method to route WS messages
+  def websocketHandler(req: AkkaHttpRequest): Flow[Message, Message, Any] = {
+    val compressFlow: Flow[TextMessage, Message, NotUsed] = chosenCompression(req.requestHeaders) match {
+      case Some(compressOpt) =>
+        Flow[TextMessage].map { tx =>
+          ByteString(tx.getStrictText)
+        }
+        .via(compressOpt.flow)
+        .map(tx => BinaryMessage(tx))
+      case None =>
+        Flow[Message]
+    }
+
+    Flow[Message]
+      .mapAsync(1) {
+        case tm: TextMessage =>
+          for {
+            input <- tryWrap(inputHandler(authHolder, tm))
+            result <- tryWrap(businessLogic(input))
+          } yield {
+            responseHandler(result)
+          }
+      }
+      .viaMat(compressFlow)(Keep.left)
+      .withAttributes(ActorAttributes.supervisionStrategy(onError))
+      .watchTermination() { (_, done) =>
+        done.map { _ =>
+          close(authHolder)
+        }
+      }
+  }
+
+  private def close(authInfo: A): Unit =
+    if (!closed.getAndSet(true))
+      onClose(authInfo)
+
+  private def tryWrap[T](input: Future[T]): Future[T] =
+    Try(input).recover({ case err: Throwable =>
+      Future.failed[T](err) }).get
```

Comment:
This does nothing. Scala's uppercase Try will only catch non-fatal exceptions, fatal exceptions won't be recovered here and are still thrown. Likewise Futures execute tryable code, hence why we match on the same Success/Failure types onComplete. So if input is a Future type then any code executing within the input future that runs without exception will resolve Success unless explicitly failed due to internal conditions. If there is a **non-fatal** exception within input then the resulting future will simply resolve as a failed future (which doesn't conform to your type bounds here anyways, scala futures aren't full pure monads). Because such an exception already resolves to a failed future then this code is just redundant

Reply:
Thanks for the clarification on intent for consumers who pass in an inputHandler or businessLogicHandler that throws before entering a Future block. Definitely a good concern, however this code won't run as written. "input" param is just a regular param so it's already evaluated and either thrown or not before being passed to tryWrap. Gotta switch to a pass by name param to delay potential throwing until you're inside the Try() bit.

Reply:
(Also if you wanted to be lazy you could just wrap the inputs manually like:
```for {
       safeInputHandler <- Future(inputHandler(...))
       input <- safeinputHandler
       safeBLogic <- Future(businessLogic(...))
      result <- safeBLogic
```
But @Kraagen already correctly pointed out to me earlier today in another repo that this introduces a new thread so for this production framework we should probably go with your separate method.